oximeter/db/src/oxql/ast/grammar.rs

```diff
@@ -0,0 +1,1125 @@
+        /// style, is supported. For example, `\u{1234}` is accepted and mapped
+        /// to `ሴ` upon parsing. This also allows users to write both quote
+        /// styles if required, by writing them as their Unicode escape
+        /// sequences. For example, this string:
+        ///
+        /// ```ignore
+        /// "this string has \u{22} in it"
+        /// ```
+        ///
+        /// Will be parsed as `this string has " in it`.
+        pub rule string_literal() -> Literal
+            = s:string_literal_impl() { Literal::String(s) }
+
+        pub(in super) rule integer_literal_impl() -> i128
+            = n:$("-"? ['0'..='9']+ !['e' | 'E' | '.'])
+        {?
+            let Ok(x) = n.parse() else {
+                return Err("integer literal");
+            };
+            if x < i128::from(i64::MIN) {
+                Err("negative overflow")
+            } else if x > i128::from(u64::MAX) {
+                Err("positive overflow")
+            } else {
+                Ok(x)
+            }
+        }
+
+        /// Parse integer literals.
+        pub rule integer_literal() -> Literal
+            = n:integer_literal_impl() { Literal::Integer(n) }
+
+        // We're being a bit lazy here, since the rule expression isn't exactly
+        // right. But we rely on calling `f64`'s `FromStr` implementation to
+        // actually verify the values can be parsed.
+        pub(in super) rule double_literal_impl() -> f64
+            = n:$("-"? ['0'..='9']* "."? ['0'..='9']* (['e' | 'E'] "-"?  ['0'..='9']+)*) {?
+                n.parse().or(Err("double literal"))
+            }
+
+        // Parse double literals.
+        pub rule double_literal() -> Literal
+            = d:double_literal_impl() { Literal::Double(d) }
+
+        /// Parse a literal.
+        ///
+        /// Literals are typed, with support for bools, durations, integers and
+        /// doubles, UUIDs, and general strings. See the rules for each type of
+        /// literal for details on supported formats.
+        pub rule literal() -> Literal
+            = lit:(
+                boolean_literal() /
+                duration_literal() /
+                integer_literal() /
+                double_literal() /
+                uuid_literal() /
+                ip_literal() /
+                string_literal() /
+                timestamp_literal()
+            )
+        {
+            lit
+        }
+
+        pub(in super) rule logical_op_impl() -> LogicalOp
+            = "||" { LogicalOp::Or} / "&&" { LogicalOp::And }
+
+        // Parse a filter item, which is one logical expression used in a filter
+        // operation.
+        #[cache_left_rec]
+        pub(in super) rule filter_item() -> Filter = precedence! {
+            // Note: We need to separate the logical operations into different
+            // levels of precedence.
+            left:(@) _? "||"  _? right:@ {
```

Comment:
Ha, it took me a while to figure out what space you were referring to! I removed it, but just for clarity it was not load-bearing. I think `cargo fmt` leaves this whole block alone since it's a massive macro call, unfortunately.

Reply:
Done in aefa41886